File: app/src/main/java/com/alexkn/syntact/domain/repository/SolvableItemRepository.java
package com.alexkn.syntact.domain.repository;

import android.content.Context;
import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.work.Data;
import androidx.work.ExistingWorkPolicy;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkManager;

import com.alexkn.syntact.data.common.AppDatabase;
import com.alexkn.syntact.data.dao.BucketDao;
import com.alexkn.syntact.data.dao.ClueDao;
import com.alexkn.syntact.data.dao.SolvableItemDao;
import com.alexkn.syntact.data.model.Bucket;
import com.alexkn.syntact.data.model.SolvableItem;
import com.alexkn.syntact.data.model.cto.SolvableTranslationCto;
import com.alexkn.syntact.domain.worker.FetchPhrasesWorker;
import com.alexkn.syntact.restservice.SolvableItemService;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Maybe;

@Singleton
public class SolvableItemRepository {

    private static final String TAG = SolvableItemRepository.class.getSimpleName();

    private final ClueDao clueDao;

    private SolvableItemService solvableItemService;

    private SolvableItemDao solvableItemDao;

    private BucketDao bucketDao;

    @Inject
    SolvableItemRepository(SolvableItemService solvableItemService, Context context) {

        this.solvableItemService = solvableItemService;
        solvableItemDao = AppDatabase.Companion.getDatabase(context).solvableItemDao();
        bucketDao = AppDatabase.Companion.getDatabase(context).bucketDao();
        clueDao = AppDatabase.Companion.getDatabase(context).clueDao();
    }

    public LiveData<List<SolvableTranslationCto>> getSolvableTranslations(Long bucketId) {

        return solvableItemDao.getSolvableTranslationsDueBefore(bucketId, Instant.now());
    }

    public Maybe<SolvableTranslationCto[]> getNextSolvableTranslations(Long bucketId, Instant time, int count) {

        Maybe<SolvableTranslationCto[]> translations = solvableItemDao.getNextTranslationDueBefore(bucketId, time, count);

        return translations.flatMap(items -> {
            if (items.length == count) {
                Log.i(TAG, "All translations present");
                return Maybe.just(items);
            } else {
                Log.i(TAG, "Translations missing, fetching remote");
                Bucket bucket = bucketDao.find(bucketId);
                long maxId = solvableItemDao.getMaxId(bucketId);
                List<SolvableTranslationCto> solvableTranslationCtos = solvableItemService.fetchNewTranslations(bucket, time, maxId, count);
                solvableTranslationCtos.forEach(cto -> {
                    solvableItemDao.insert(cto.getSolvableItem());
                    clueDao.insert(cto.getClue());
                });
                //TODO Service to return Maybe
                return Maybe.just(solvableTranslationCtos.toArray(new SolvableTranslationCto[0]));
            }
        });
    }

    public void fetchSolvableItems(long bucketId, Instant startTime) {

        Data data = new Data.Builder().putLong("bucketId", bucketId).putLong("timestamp", startTime.toEpochMilli()).build();

        OneTimeWorkRequest workRequest = new OneTimeWorkRequest.Builder(FetchPhrasesWorker.class).setInputData(data).build();
        WorkManager.getInstance().enqueueUniqueWork(FetchPhrasesWorker.class.getName(), ExistingWorkPolicy.KEEP, workRequest);
    }

    public void solvePhrase(SolvableTranslationCto solvableTranslation) {

        SolvableItem solvableItem = solvableTranslation.getSolvableItem();
        Log.i(TAG, "Solving Phrase " + solvableItem.getText());

        int performanceRating = 3;
        float easiness = solvableItem.getEasiness();
        easiness += -0.80 + 0.28 * performanceRating + 0.02 * performanceRating * performanceRating;

        int consecutiveCorrectAnswers = solvableItem.getConsecutiveCorrectAnswers() + 1;

        double daysToAdd = 6 * Math.pow(easiness, consecutiveCorrectAnswers - 1d);
        Instant nextDueDate = Instant.now().plus(Math.round(daysToAdd), ChronoUnit.DAYS);

        solvableItem.setEasiness(easiness);
        solvableItem.setConsecutiveCorrectAnswers(consecutiveCorrectAnswers);
        solvableItem.setNextDueDate(nextDueDate);
        solvableItem.setLastSolved(Instant.now());
        solvableItem.setTimesSolved(solvableItem.getTimesSolved() + 1);

        solvableItemDao.update(solvableItem);
    }
}

File: app/src/main/java/com/alexkn/syntact/domain/repository/SolvableItemRepository.kt
package com.alexkn.syntact.domain.repository

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.work.*
import com.alexkn.syntact.app.Property
import com.alexkn.syntact.app.TAG
import com.alexkn.syntact.data.dao.ClueDao
import com.alexkn.syntact.data.dao.SolvableItemDao
import com.alexkn.syntact.data.model.Clue
import com.alexkn.syntact.data.model.cto.SolvableTranslationCto
import com.alexkn.syntact.domain.worker.FetchPhrasesWorker
import com.alexkn.syntact.rest.service.SyntactService
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.lang.Exception
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.max
import kotlin.math.pow
import kotlin.math.roundToLong

@Singleton
class SolvableItemRepository @Inject
internal constructor(
        private val property: Property,
        private val syntactService: SyntactService,
        private val solvableItemDao: SolvableItemDao,
        private val clueDao: ClueDao
) {

    fun getSolvableTranslations(bucketId: Long): LiveData<List<SolvableTranslationCto>> = solvableItemDao.getSolvableTranslations(bucketId)

    suspend fun findNextSolvableTranslation(bucketId: Long, time: Instant): SolvableTranslationCto {

        val nextTranslation = solvableItemDao.getNextTranslationDueBefore(bucketId, time)

        return if (nextTranslation.clue == null) {
            val token = "Token " + property["api-auth-token"]
            val translations = syntactService.getTranslations(token, nextTranslation.solvableItem.translationUrl, Locale.getDefault().language)
            if (translations.size > 1) {
                Log.i(TAG, "Multiple Translations not yet supported, using first translation and discarding others")
            }
            if (translations.isEmpty()) {
                throw Exception("No TranslationResponse found for ${nextTranslation.solvableItem}")
            }
            val translation = translations[0]
            val clue = Clue(
                    id = translation.id,
                    text = translation.text.capitalize(),
                    solvableItemId = nextTranslation.solvableItem.id
            )
            clueDao.insert(clue)
            Log.i(TAG, "Saved Clue $clue")
            solvableItemDao.getNextTranslationDueBefore(bucketId, time)
        } else {
            nextTranslation
        }

    }

    fun fetchSolvableItems(bucketId: Long, count: Int? = null) {

        val data = Data.Builder().putLong("bucketId", bucketId)
        count?.let { data.putInt("count", count) }
        val constraints = Constraints.Builder().setRequiredNetworkType(NetworkType.CONNECTED).build()
        val workRequest = OneTimeWorkRequest.Builder(FetchPhrasesWorker::class.java).setInputData(data.build()).setConstraints(constraints).build()
        WorkManager.getInstance().enqueueUniqueWork(FetchPhrasesWorker::class.java.name, ExistingWorkPolicy.KEEP, workRequest)
    }

    suspend fun disableSolvableItem(solvableTranslationCto: SolvableTranslationCto) {

        val solvableItem = solvableTranslationCto.solvableItem
        solvableItem.disabled = true;
        solvableItemDao.update(solvableItem)
    }

    suspend fun solvePhrase(solvableTranslation: SolvableTranslationCto) {

        val solvableItem = solvableTranslation.solvableItem
        Log.i(TAG, "Solving PhraseResponse " + solvableItem.text)

        val performanceRating = 3
        var easiness = solvableItem.easiness
        easiness += (-0.80 + 0.28 * performanceRating + 0.02 * performanceRating.toDouble() * performanceRating.toDouble()).toFloat()

        val consecutiveCorrectAnswers = solvableItem.consecutiveCorrectAnswers + 1

        val daysToAdd = 6 * easiness.toDouble().pow(consecutiveCorrectAnswers - 1.0)
        val nextDueDate = Instant.now().plus(daysToAdd.roundToLong(), ChronoUnit.DAYS)

        solvableItem.easiness = easiness
        solvableItem.consecutiveCorrectAnswers = consecutiveCorrectAnswers
        solvableItem.nextDueDate = nextDueDate
        solvableItem.lastSolved = Instant.now()
        solvableItem.timesSolved = solvableItem.timesSolved + 1

        solvableItemDao.update(solvableItem)
    }

    suspend fun phraseIncorrect(solvableTranslation: SolvableTranslationCto) {

        val solvableItem = solvableTranslation.solvableItem
        Log.i(TAG, "PhraseResponse incorrect " + solvableItem.text)

        val performanceRating = 1
        var easiness = solvableItem.easiness
        easiness += (-0.80 + 0.28 * performanceRating + 0.02 * performanceRating.toDouble() * performanceRating.toDouble()).toFloat()
        easiness = max(easiness, 1.3f)

        val consecutiveCorrectAnswers = 0

        val nextDueDate = Instant.now().plus(1, ChronoUnit.DAYS)

        solvableItem.easiness = easiness
        solvableItem.consecutiveCorrectAnswers = consecutiveCorrectAnswers
        solvableItem.nextDueDate = nextDueDate

        solvableItemDao.update(solvableItem)
    }

}
